core/src/modules/ConfigOutput.cpp

```diff
@@ -51,35 +66,93 @@ void ConfigOutput::configure()
+    if (currentFileName == "") {
+        std::stringstream startStream;
+        startStream << meta.time();
+        currentFileName = m_filePrefix + ".nc";
+    }
 
     ModelState state;
-    const ModelState* pState;
     if (outputAllTheFields) {
-        pState = &fullState;
+        for (const auto& entry : protectedArrayNames) {
+            ModelArrayConstReference macr
+                = getProtectedArray().at(static_cast<size_t>(entry.second));
+            if (macr && macr->trueSize() > 0)
+                state.data[entry.first] = *macr;
+        }
+        for (const auto& entry : sharedArrayNames) {
+            ModelArrayReference mar = getSharedArray().at(static_cast<size_t>(entry.second));
+            if (mar && mar->trueSize() > 0)
+                state.data[entry.first] = *mar;
+        }
     } else {
         // Filter only the given fields to the output state
-        for (auto fieldEntry : fullState.data) {
-            if (fieldsForOutput.count(fieldEntry.first) > 0) {
-                state.data[fieldEntry.first] = fieldEntry.second;
-            }
+        for (const auto& fieldExtName : fieldsForOutput) {
+            if (protectedExternalNames.count(fieldExtName)) {
+                ModelArrayConstReference macr = getProtectedArray().at(static_cast<size_t>(
+                    protectedArrayNames.at(protectedExternalNames.at(fieldExtName))));
+                if (macr)
+                    state.data[fieldExtName] = *macr;
+            } else if (sharedExternalNames.count(fieldExtName)) {
+                ModelArrayReference mar = getSharedArray().at(
+                    static_cast<size_t>(sharedArrayNames.at(sharedExternalNames.at(fieldExtName))));
+                if (mar)
+                    state.data[fieldExtName] = *mar;
+            } // else do not add any data to the state under that name
         }
-        pState = &state;
     }
 
-    Logged::info("ConfigOutput: Outputting " + std::to_string(pState->data.size()) + " fields to "
-        + timeFileName + "\n");
-
-    if ((everyTS && meta.time() >= lastOutput) || (meta.time() >= lastOutput + outputPeriod)) {
-        StructureFactory::fileFromState(*pState, meta, timeFileName);
+    /*
+     * Produce output either:
+     *    • on every timestep after the start time initially stored in lastOutput
+     *    • whenever the current time is an integer number of time periods from the
+     *      last output time.
+     */
+    if ((everyTS && meta.time() >= lastOutput)
+        || (std::fmod((meta.time() - lastOutput).seconds(), outputPeriod.seconds()) == 0.)) {
+        Logged::info("ConfigOutput: Outputting " + std::to_string(state.data.size()) + " fields to "
+            + currentFileName + " at " + meta.time().format() + "\n");
+        StructureFactory::fileFromState(state, meta, currentFileName, false);
         lastOutput = meta.time();
     }
 }
+
+std::string concatenateFields(const std::set<std::string>& strSet)
```

Comment:
This will leave a dangling , comma in the output. Is that okay? (Probably it is but just wanted to check!).